circuits/src/zk_circuits/valid_commitments.rs

```diff
@@ -198,98 +195,97 @@ where
             .iter()
             .zip(augmented_wallet.orders.iter())
         {
-            EqGadget::constrain_eq(base_order.clone(), augmented_order.clone(), cs);
+            EqGadget::constrain_eq(base_order, augmented_order, cs)?;
         }
 
         // All fees should be the same
         for (base_fee, augmented_fee) in base_wallet.fees.iter().zip(augmented_wallet.fees.iter()) {
-            EqGadget::constrain_eq(base_fee.clone(), augmented_fee.clone(), cs);
+            EqGadget::constrain_eq(base_fee, augmented_fee, cs)?;
         }
 
         // Keys should be equal
-        EqGadget::constrain_eq(base_wallet.keys.clone(), augmented_wallet.keys.clone(), cs);
+        EqGadget::constrain_eq(&base_wallet.keys, &augmented_wallet.keys, cs)?;
 
         // Blinders should be equal
-        EqGadget::constrain_eq(
-            base_wallet.blinder.clone(),
-            augmented_wallet.blinder.clone(),
-            cs,
-        );
+        EqGadget::constrain_eq(&base_wallet.blinder, &augmented_wallet.blinder, cs)
     }
 
     /// Verify that the wallet has the given balance at the specified index
-    fn contains_balance_at_index<CS: RandomizableConstraintSystem>(
+    fn contains_balance_at_index(
         index: Variable,
-        target_balance: BalanceVar<Variable>,
-        wallet: &WalletVar<LinearCombination, MAX_BALANCES, MAX_ORDERS, MAX_FEES>,
-        cs: &mut CS,
-    ) {
-        let mut curr_index: LinearCombination = Variable::Zero().into();
-        let mut balance_found: LinearCombination = Variable::Zero().into();
+        target_balance: &BalanceVar,
+        wallet: &WalletVar<MAX_BALANCES, MAX_ORDERS, MAX_FEES>,
+        cs: &mut PlonkCircuit,
+    ) -> Result<(), CircuitError> {
+        let mut curr_index = cs.zero();
+        let mut balance_found = cs.false_var();
+
         for balance in wallet.balances.iter() {
-            let index_mask =
-                EqGadget::eq::<LinearCombination, Variable, _, _, _>(curr_index.clone(), index, cs);
-            let balances_eq = EqGadget::eq(balance.clone(), target_balance.clone(), cs);
+            let index_mask = EqGadget::eq(&curr_index, &index, cs)?;
+            let balances_eq = EqGadget::eq(balance, target_balance, cs)?;
+
+            let found = cs.logic_and(index_mask, balances_eq)?;
 
-            let found = AndGate::and(index_mask, balances_eq, cs);
-            balance_found += found;
-            curr_index += Variable::One();
+            balance_found = cs.logic_or(balance_found, found)?;
```

Comment:
Using `logic_or` here feels more readable / semantically appropriate, but doesn't it make it such that we no longer constrain that a *single* balance is found? I.e. previously, if somehow two balances were matched, `balance_found = 2 =\= 1` would not satisfy the constraint, but now, `true || true == true` does satisfy

Reply:
the index mask makes these equivalent, the weird one is fees which i'll note in the comments.